### PR TITLE
[Do Not Review][Terrarium - Providers]: Implementation of Upload Feature - 1

### PR DESCRIPTION
**Pull request by anjanikshree12**
### This PR includes:
* upload feature support being implemented as a part of CIEDEV-3780
* Protobuff definitions for upload service having functionalities such as upload provider binary, shasum and signature files for the storage and gateway.
* gRPC code generated by protoc compiler tool from the protobuff definitions.

> [!NOTE]  
> ProviderRequest struct has been shifted from storage.proto to provider.proto & corresponding changes has been done accordingly to support this shift.

> [!IMPORTANT]  
>In the upcoming PRs as a part of same feature support, the missing methods will be added.(current build failing) and this PR will be the base PR to remaining ones.

### CODE REVIEW

File: pb/terrarium/provider/provider.proto [hunk 1]
```diff
@@ -4,6 +4,9 @@ package terrarium.provider;
 option go_package = "github.com/terrariumcloud/terrarium/pkg/terrarium/provider";
 
 service ProviderPublisher {
+  rpc UploadProviderSourceZip (stream UploadProviderSourceZipRequest) returns (Response) {}
```
Comment (odelanno) [inline pb/terrarium/provider/provider.proto:7]: is the plan to upload source file? it does not make sense here as we don't want to build provider in terrarium. the function should clearly describe the purpose

**Review by odelanno — CHANGES_REQUESTED** [review]

File: pb/terrarium/provider/provider.proto [hunk 2]
```diff
@@ -84,11 +84,11 @@ message ProviderRequest {
   string arch = 4;
 }
 
-message UploadProviderSourceZipRequest {
+message UploadProviderBinaryRequest {
   ProviderRequest provider = 1;
```
Comment (odelanno) [inline pb/terrarium/provider/provider.proto:88]: How will the request identify the provider arch / os combination the provider request already contains a list of repeated platform items. that will not enable you to specify how a specific upload is for a given os / arch. How are you planning to map binary with an architecture?

File: pb/terrarium/provider/provider.proto [hunk 3]
```diff
@@ -84,11 +84,11 @@ message ProviderRequest {
...
   bytes zip_data_chunk = 2;
```
Comment (odelanno) [inline pb/terrarium/provider/provider.proto:89]: does it needs to be a zip ? or will it be a raw binary ?

**Review by odelanno — CHANGES_REQUESTED** [review]

File: pb/terrarium/provider/provider.proto (hunk 1 above)
Comment (anjanikshree12) [reply pb/terrarium/provider/provider.proto]: No, source file won't be uploaded, the plan is that the provider will be build as a part of our workflows in some runners, **this function's role is to upload the provider binary.**
I agree to your point that the function name should be more relatable and updated it accordingly.

File: pb/terrarium/provider/provider.proto (hunk 3 above)
Comment (anjanikshree12) [reply pb/terrarium/provider/provider.proto:89]: It needs to be a zip.

**Review by odelanno — APPROVED** [review]

**Pull request closed without merging**